stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp

```diff
@@ -67,31 +75,180 @@ namespace stan {
+                                    c[y[i]-1] - lambda[i]);
+        }
+      }
+      return logp_n;
+    }
 
-      // log(1 - inv_logit(lambda))
-      if (y == 1)
-        return -log1p_exp(lambda - c(0));
+    template <typename T_loc, typename T_cut>
+    typename boost::math::tools::promote_args<T_loc, T_cut>::type
+    ordered_logistic_lpmf(const std::vector<int>& y,
+                      const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
+                      const Eigen::Matrix<T_cut, Eigen::Dynamic, 1>& c) {
+      return ordered_logistic_lpmf<false>(y, lambda, c);
+    }
 
-      // log(inv_logit(lambda - c(K-3)));
-      if (y == K) {
-        return -log1p_exp(c(K-2) - lambda);
+    /**
+     * Returns the (natural) log probability of the specified array 
+     * of integers given the vector of continuous locations and 
+     * array of specified cutpoints in an ordered logistic model.
+     *
+     * <p>Typically the continous location
+     * will be the dot product of a vector of regression coefficients
+     * and a vector of predictors for the outcome.
+     *
+     * @tparam propto True if calculating up to a proportion.
+     * @tparam T_y Type of y variable (should be std::vector<int>).
+     * @tparam T_loc Location type.
+     * @tparam T_cut Cut-point type.
+     * @param y Array of integers
+     * @param lambda Vector of continuous location variables.
+     * @param c array of Positive increasing vectors of cutpoints.
+     * @return Log probability of outcome given location and
+     * cutpoints.
+
+     * @throw std::domain_error If the outcome is not between 1 and
+     * the number of cutpoints plus 2; if the cutpoint vector is
+     * empty; if the cutpoint vector contains a non-positive,
+     * non-finite value; or if the cutpoint vector is not sorted in
+     * ascending order.
+     * @throw std::invalid_argument If y and lambda are different
+     * lengths, or if y and the array of cutpoints are of different
+     * lengths.
+     */
+    template <bool propto, typename T_loc, typename T_cut>
+    typename boost::math::tools::promote_args<T_loc, T_cut>::type
+    ordered_logistic_lpmf(const std::vector<int>& y,
+                      const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
+                      const std::vector<Eigen::Matrix<
+                                              T_cut, Eigen::Dynamic, 1>>& c) {
+      typename boost::math::tools::promote_args<T_loc, T_cut>::type logp_n(0.0);
+
+      using boost::math::tools::promote_args;
+      using std::exp;
+      using std::log;
+
+      static const std::string function = "ordered_logistic";
+
+      int N = lambda.size();
+
+      check_consistent_sizes(function, "Integers", y, "Locations", lambda);
+      check_consistent_sizes(function, "Integers", y, "Cut-points", c);
+
+      for (int i = 0; i < N; ++i) {
+        int K = c[i].size() + 1;
+
+        check_bounded(function, "Random variable", y[i], 1, K);
```

Comment:
These should still be outside of the loop so that you can use vectorized version.  the vectorized ones give more sensible error messages because they include the index.  here, we don't want to put it in the message because it takes too long to do all the string fiddling---it has to be done inside the test.